library/core/src/str/validations/utf8_dfa.rs

```diff
@@ -0,0 +1,324 @@
+                    Some(if pos == valid_up_to { 1 } else { (pos - valid_up_to) as u8 });
+                return Err(Utf8Error { valid_up_to, error_len });
+            }
+            // Keep going
+            _ => {}
+        }
+        pos += 1;
+    }
+    if state != END { Err(Utf8Error { valid_up_to, error_len: None }) } else { Ok(()) }
+}
+
+#[inline]
+fn validate_utf8_impl(inp: &[u8]) -> Result<(), Utf8Error> {
+    let mut state = END;
+    let mut pos = 0;
+    let (chunks, tail) = inp.as_chunks::<CHUNK_LEN>();
+    if !chunks.is_empty() {
+        let mut last_state = state;
+        let mut chunk_iter = chunks.iter();
+        while let Some(chunk) = chunk_iter.next() {
+            if state == END && all_ascii_chunk(chunk) {
+                let skipped = skip_ascii_chunks(&mut chunk_iter);
+                pos += skipped * CHUNK_LEN;
+            } else {
+                state = dfa_run(state, chunk);
+                if unlikely(state == ERR) {
+                    break;
+                }
+                last_state = state;
+            }
+            pos += core::mem::size_of_val(chunk);
+        }
+        if unlikely(state == ERR) {
+            return Err(utf8_find_error(inp, pos, last_state != END));
+        }
+    }
+    // Did we leave the optimized loop in the middle of a UTF-8 sequence?
+    let was_mid_char = state != END;
+    debug_assert!(state != ERR);
+    if !tail.is_empty() {
+        // Check and early return if the last CHUNK_LEN bytes were all ASCII. The
+        // motivation here is to avoid bringing the DFA table into the cache for
+        // pure ASCII.
+        //
+        // 1. avoid touching the DFA table for pure-ASCII input.
+        // 2. add a branch into the `dfa_run` inner loop.
+        //
+        // So we check and see if the last CHUNK_LEN bytes were all ASCII. This does
+        // compare with a few bytes that we've already processed, but handling
+        // that is not required for correctness (and doing so seems ot hurt
+        // performance)
+        if !was_mid_char && inp.len() >= CHUNK_LEN && tail.len() < CHUNK_LEN {
+            use crate::convert::TryFrom;
+            let range = (inp.len() - CHUNK_LEN)..inp.len();
+            debug_assert!(range.contains(&pos), "{:?}", (range, pos));
+            if all_ascii_chunk(<&[u8; CHUNK_LEN]>::try_from(&inp[range]).unwrap()) {
+                return Ok(());
+            }
+        }
+
+        state = dfa_run(state, tail);
+    }
+
+    if likely(state == END) {
+        return Ok(());
+    }
+    let (index, backup) =
+        if state == ERR { (inp.len() - tail.len(), was_mid_char) } else { (inp.len(), true) };
+    Err(utf8_find_error(inp, index, backup))
+}
+
+#[inline]
+fn backup_not_yet_invalid(inp: &[u8], mut pos: usize) -> usize {
+    debug_assert!(!inp.is_empty() && inp.get(..pos).is_some());
+    while pos != 0 {
+        pos -= 1;
+        let is_cont = (inp[pos] & 0b1100_0000) == 0b1000_0000;
```

Comment:
Well, this isn't hot code so probably not significant, but the `as i8 >= -64` construction from the old impl should take 1 less instruction since it can use status flags for signed operations for that comparison.

Reply:
Yeah good point. I think there's a function in the parent module too.